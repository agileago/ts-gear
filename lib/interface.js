"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./tool/types");
exports.httpMethods = types_1.tuple('get', 'put', 'post', 'delete', 'options', 'head', 'patch');
