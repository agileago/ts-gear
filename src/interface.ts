import { JSONSchema4 } from 'json-schema'

export type JSONSchema = JSONSchema4

/** json schema每个节点traverse结构 */
export interface ISchemaNode {
  value: any
  key: string
  parent: any
  path: string[]
}

/** 每个请求参数的原始schema结构 */
export interface IParameter {
  name: string
  in: keyof IParameterSchema
  description: string
  required: boolean
  type: string
  items?: JSONSchema
  format?: string
  schema?: JSONSchema
}

// interface IResponseSchema extends JSONSchema {
//   title?: string
// }

interface IResponse {
  // '200'?: {
  //   description: string
  //   schema: IResponseSchema
  // }
  [key: string]: {
    description: string
    schema?: JSONSchema
  }
}

/** swagger paths内每个http请求详情对应的结构 */
export interface IRequestDetail {
  parameters?: IParameter[]
  summary?: string
  description?: string
  operationId?: string
  produces: string[]
  tags: string[]
  responses: IResponse
  security?: any
  deprecated: boolean
}

interface IRequest {
  /** http method */
  [path: string]: IRequestDetail
}

/** swagger "paths" */
export interface IPaths {
  /** api url path */
  [k: string]: IRequest
}

/** 每个请求的json schema定义的parameters的住装数据结构 */
export interface IParameterSchema {
  query?: JSONSchema
  body?: JSONSchema
  path?: JSONSchema
  formData?: JSONSchema
  header?: JSONSchema
}

export type TPathMatcherFunction = RegExp | ((url: string) => boolean)

export interface IProject {
  /** 项目名字，会在dest文件夹中创建该项目名称的目录 */
  name: string
  /** 读取swagger配置的路径，可以是本地json文件或远程swagger地址
   * http开头的为远程地址
   * 否则按本地文件读取
   * */
  source: string
  /** 需要验证参数在这里添加 */
  fetchOption?: RequestInit
  /** 过滤需要的请求路径
   * 有些项目其中掺杂了很多无用的api，过滤掉可减少文件体积，加速运行
   * */
  pathMatcher?: TPathMatcherFunction
}

/** 用户配置文件定义 */
export interface IUserConfig {
  /** 项目输出文件夹，会在该文件夹下建立以每个项目名建立独立的文件夹
   * 默认为 './service'
   * */
  dest: string

  projects: IProject[]
}
