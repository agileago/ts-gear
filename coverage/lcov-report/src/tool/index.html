
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/tool</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/tool</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>182/182</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">99.07% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>106/107</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>35/35</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>175/175</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="assembleDoc.ts"><a href="assembleDoc.ts.html">assembleDoc.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="21" class="abs high">21/21</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="18" class="abs high">18/18</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="21" class="abs high">21/21</td>
	</tr>

<tr>
	<td class="file high" data-value="camelCase.ts"><a href="camelCase.ts.html">camelCase.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="cleanName.ts"><a href="cleanName.ts.html">cleanName.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="13" class="abs high">13/13</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="15" class="abs high">15/15</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="13" class="abs high">13/13</td>
	</tr>

<tr>
	<td class="file high" data-value="clearObject.ts"><a href="clearObject.ts.html">clearObject.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="genericType.ts"><a href="genericType.ts.html">genericType.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="46" class="abs high">46/46</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="16" class="abs high">16/16</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="42" class="abs high">42/42</td>
	</tr>

<tr>
	<td class="file high" data-value="getDefinition.ts"><a href="getDefinition.ts.html">getDefinition.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

<tr>
	<td class="file high" data-value="isReference.ts"><a href="isReference.ts.html">isReference.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

<tr>
	<td class="file high" data-value="log.ts"><a href="log.ts.html">log.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="patchGlobalDefinitionMap.ts"><a href="patchGlobalDefinitionMap.ts.html">patchGlobalDefinitionMap.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="prettierWrite.ts"><a href="prettierWrite.ts.html">prettierWrite.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	</tr>

<tr>
	<td class="file high" data-value="schemaToTypescript.ts"><a href="schemaToTypescript.ts.html">schemaToTypescript.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="46" class="abs high">46/46</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="38" class="abs high">38/38</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="43" class="abs high">43/43</td>
	</tr>

<tr>
	<td class="file high" data-value="transformSwaggerPathToRouterPath.ts"><a href="transformSwaggerPathToRouterPath.ts.html">transformSwaggerPathToRouterPath.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="translate.ts"><a href="translate.ts.html">translate.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="traverseSchema.ts"><a href="traverseSchema.ts.html">traverseSchema.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="11" class="abs high">10/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Thu Apr 23 2020 19:25:35 GMT+0800 (GMT+08:00)
            </div>
        </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    