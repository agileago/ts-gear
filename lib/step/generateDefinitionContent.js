"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ts_morph_1 = require("ts-morph");
const schemaToTypescript_1 = require("../tool/schemaToTypescript");
const source_1 = require("../source");
const projectGlobalVariable_1 = require("../projectGlobalVariable");
const assembleDoc_1 = require("../tool/assembleDoc");
/** generate on definition ts content */
exports.generateDefinitionContent = (project) => {
    const { definitionMap } = projectGlobalVariable_1.getGlobal(project);
    Object.values(definitionMap).forEach((definition) => {
        if (definition.typescriptContent || !definition.schema) {
            return;
        }
        const source = source_1.sow();
        const title = definition.typeName;
        const schema = definition.schema;
        if (schema.type === 'object') {
            if (schema.properties) {
                const preferInterface = Boolean(project.preferInterface);
                const declarationOptin = {
                    isExported: true,
                    name: title,
                    typeParameters: definition.typeParameters
                        ? definition.typeParameters.map((t) => ({
                            name: t,
                            default: 'any',
                        }))
                        : undefined,
                    docs: assembleDoc_1.assembleDoc(schema),
                };
                const klass = preferInterface ? source.addInterface(declarationOptin) : source.addClass(declarationOptin);
                Object.getOwnPropertyNames(schema.properties).forEach((name) => {
                    const property = schema.properties[name];
                    const propertyStructure = {
                        name,
                        type: schemaToTypescript_1.schemaToTypescript(property),
                        scope: preferInterface ? undefined : ts_morph_1.Scope.Public,
                        hasQuestionToken: !schema.required || !schema.required.includes(name),
                        docs: assembleDoc_1.assembleDoc(property),
                    };
                    /** interface property can not has default value
                      so use class as type */
                    if (!preferInterface && Reflect.has(property, 'default')) {
                        propertyStructure.initializer = String(property.default);
                    }
                    klass.addProperty(propertyStructure);
                });
                // 没有properties，会有additionalProperties
            }
            else if (schema.additionalProperties) {
                const { additionalProperties } = schema;
                // class doesn`t has "addIndexSignature", so use interface
                source.addInterface({
                    isExported: true,
                    name: title,
                    docs: typeof additionalProperties !== 'boolean' ? assembleDoc_1.assembleDoc(additionalProperties) : [],
                    indexSignatures: [
                        {
                            keyName: 'key',
                            keyType: 'string',
                            returnType: additionalProperties === true ? 'any' : schemaToTypescript_1.schemaToTypescript(additionalProperties),
                        },
                    ],
                });
            }
            else {
                source.addTypeAlias({
                    isExported: true,
                    name: title,
                    type: schemaToTypescript_1.schemaToTypescript(schema),
                    docs: assembleDoc_1.assembleDoc(schema),
                });
            }
        }
        else {
            source.addTypeAlias({
                isExported: true,
                name: title,
                type: schemaToTypescript_1.schemaToTypescript(schema),
                docs: assembleDoc_1.assembleDoc(schema),
            });
        }
        definition.typescriptContent = source_1.harvest(source);
    });
};
