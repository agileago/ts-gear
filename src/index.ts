export { requester as fetchRequester } from './requester/fetch'
export { requester as axiosRequester } from './requester/axios'
export { IProject, PropertyType } from './interface'
