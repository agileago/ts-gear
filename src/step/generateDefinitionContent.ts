import {
  OptionalKind,
  PropertyDeclarationStructure,
  Scope,
  PropertySignatureStructure,
  InterfaceDeclarationStructure,
  ClassDeclarationStructure,
} from 'ts-morph'

import { schemaToTypescript } from '../tool/schemaToTypescript'
import { sow, harvest } from '../source'
import { IProject } from '../interface'
import { getGlobal } from '../projectGlobalVariable'
import { assembleDoc } from '../tool/assembleDoc'

/** generate on definition ts content */
export const generateDefinitionContent = (project: IProject) => {
  const { definitionMap } = getGlobal(project)
  Object.values(definitionMap).forEach((definition) => {
    if (definition.typescriptContent || !definition.schema) {
      return
    }
    const source = sow()
    const title = definition.typeName!
    const schema = definition.schema!
    if (schema.type === 'object') {
      if (schema.properties) {
        const preferInterface = Boolean(project.preferInterface)
        const declarationOptin: OptionalKind<InterfaceDeclarationStructure & ClassDeclarationStructure> = {
          isExported: true,
          name: title,
          typeParameters: definition.typeParameters
            ? definition.typeParameters.map((t) => ({
                name: t,
                default: 'any',
              }))
            : undefined,
          docs: assembleDoc(schema),
        }
        const klass = preferInterface ? source.addInterface(declarationOptin) : source.addClass(declarationOptin)
        Object.getOwnPropertyNames(schema.properties).forEach((name) => {
          const property = schema!.properties![name]
          const propertyStructure: OptionalKind<PropertyDeclarationStructure & PropertySignatureStructure> = {
            name,
            type: schemaToTypescript(property),
            scope: preferInterface ? undefined : Scope.Public,
            hasQuestionToken: !schema.required || !schema.required.includes(name),
            docs: assembleDoc(property),
          }
          /** interface property can not has default value
            so use class as type */
          if (!preferInterface && Reflect.has(property, 'default')) {
            propertyStructure.initializer = String(property.default)
          }
          klass.addProperty(propertyStructure)
        })
        // 没有properties，会有additionalProperties
      } else if (schema.additionalProperties) {
        const { additionalProperties } = schema
        // class doesn`t has "addIndexSignature", so use interface
        source.addInterface({
          isExported: true,
          name: title,
          docs: typeof additionalProperties !== 'boolean' ? assembleDoc(additionalProperties) : [],
          indexSignatures: [
            {
              keyName: 'key',
              keyType: 'string',
              returnType: additionalProperties === true ? 'any' : schemaToTypescript(additionalProperties),
            },
          ],
        })
      } else {
        source.addTypeAlias({
          isExported: true,
          name: title,
          type: schemaToTypescript(schema),
          docs: assembleDoc(schema),
        })
      }
    } else {
      source.addTypeAlias({
        isExported: true,
        name: title,
        type: schemaToTypescript(schema),
        docs: assembleDoc(schema),
      })
    }
    definition.typescriptContent = harvest(source)
  })
}
