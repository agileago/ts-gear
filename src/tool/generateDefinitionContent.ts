import { IndexSignatureDeclarationStructure, OptionalKind, PropertyDeclarationStructure, Scope } from 'ts-morph'

import { transformSwaggerPropertyToTsType } from './transformSwaggerPropertyToTsType'

import { sow, harvest } from 'src/source'
import { JSONSchema } from 'src/interface'
import { definitionMap } from 'src/global'
import { hasGenericSymbol } from 'src/tool/parseGenericType'

/** generate on definition ts content */
export const generateDefinitionContent = (definition: JSONSchema, title: string) => {
  const source = sow()
  if (hasGenericSymbol(title)) {
    const names = title.split(/<|>/).filter(Boolean)
    /** generac type */
    const [parentTypeName, ...typeParameters] = names
    if (!(parentTypeName in definitionMap)) {
      const klass = source.addClass({
        isExported: true,
        name: parentTypeName,
        typeParameters: typeParameters,
      })
      definitionMap[parentTypeName] = ''
      if (definition.description) {
        klass.addJsDoc(definition.description)
      }
      for (const name of Object.getOwnPropertyNames(definition.properties)) {
        const property = definition!.properties![name]
        const klassStructure: OptionalKind<PropertyDeclarationStructure> = {
          name,
          type: transformSwaggerPropertyToTsType(property),
          scope: Scope.Public,
          hasQuestionToken: !definition.required || !definition.required.includes(name),
        }
        /** interface property can not has default value
            so use class as type */
        if (Reflect.has(property, 'default')) {
          klassStructure.initializer = String(property.default)
        }
        if (Reflect.has(property, 'description')) {
          klassStructure.docs = [String(property.description)]
        }
        klass.addProperty(klassStructure)
      }
    }
  } else {
    if (definition.type === 'object') {
      if (definition.properties) {
        const klass = source.addClass({
          isExported: true,
          name: title,
        })
        if (definition.description) {
          klass.addJsDoc(definition.description)
        }
        for (const name of Object.getOwnPropertyNames(definition.properties)) {
          const property = definition!.properties![name]
          const klassStructure: OptionalKind<PropertyDeclarationStructure> = {
            name,
            type: transformSwaggerPropertyToTsType(property),
            scope: Scope.Public,
            hasQuestionToken: !definition.required || !definition.required.includes(name),
          }
          /** interface property can not has default value
            so use class as type */
          if (Reflect.has(property, 'default')) {
            klassStructure.initializer = String(property.default)
          }
          if (Reflect.has(property, 'description')) {
            klassStructure.docs = [String(property.description)]
          }
          klass.addProperty(klassStructure)
        }
        // 没有properties，会有additionalProperties
      } else if (definition.additionalProperties) {
        // class doesn`t has "addIndexSignature", so use interface
        const interFace = source.addInterface({
          isExported: true,
          name: title,
        })
        const additionalProperties = definition.additionalProperties as JSONSchema
        const interfaceStructure: OptionalKind<IndexSignatureDeclarationStructure> = {
          keyName: 'key',
          keyType: 'string',
          returnType: transformSwaggerPropertyToTsType(additionalProperties),
        }
        if (Reflect.has(additionalProperties, 'description')) {
          interfaceStructure.docs = [String(additionalProperties.description)]
        }
        interFace.addIndexSignature(interfaceStructure)
      } else {
        source.addTypeAlias({
          isExported: true,
          name: title,
          type: transformSwaggerPropertyToTsType(definition),
        })
      }
    } else {
      source.addTypeAlias({
        isExported: true,
        name: title,
        type: transformSwaggerPropertyToTsType(definition),
      })
    }
  }
  return harvest(source)
}
