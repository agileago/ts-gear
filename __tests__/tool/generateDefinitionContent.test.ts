import { generateDefinitionContent } from 'src/tool/generateDefinitionContent'
import { JSONSchema } from 'src/interface'

describe('src/tool/generateDefinitionContent', () => {
  it('generate definition class', () => {
    const apiResponse = {
      type: 'object',
      properties: {
        code: {
          type: 'integer',
          format: 'int32',
        },
        type: {
          type: 'string',
        },
        message: {
          type: 'string',
        },
      },
    }
    const res = generateDefinitionContent(apiResponse as JSONSchema, 'ApiResponse')
    expect(res).toMatchSnapshot()
  })

  it('generic type', () => {
    const apiResponse = {
      type: 'object',
      properties: {
        code: {
          type: 'integer',
          format: 'int32',
        },
        type: {
          type: 'string',
        },
        entities: {
          type: 'array',
          description: '数据列表',
          items: {
            $ref: '#/definitions/DataVO',
          },
        },
        message: {
          type: 'string',
        },
      },
    }
    const res = generateDefinitionContent(apiResponse as JSONSchema, 'ApiResponse<DataVO>')
    expect(res).toMatchSnapshot()
  })
})
