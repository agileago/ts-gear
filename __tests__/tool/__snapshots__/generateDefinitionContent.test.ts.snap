// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/tool/generateDefinitionContent generate definition class 1`] = `
"export class ApiResponse {
    public code?: number;
    public type?: string;
    public message?: string;
}
"
`;

exports[`src/tool/generateDefinitionContent generic type 1`] = `
"export class ApiResponse<DataVO> {
    public code?: number;
    public type?: string;
    /** 数据列表 */
    public entities?: Array<DataVO>;
    public message?: string;
}
"
`;
